src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -674,6 +675,76 @@ ss::future<init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
       });
 }
 
+namespace {
+
+bool is_epoch_overflow(int16_t epoch) {
+    return epoch >= std::numeric_limits<int16_t>::max();
+}
+
+bool is_valid_expected_pid(
+  const tm_transaction& tx, const model::producer_identity& expected_pid) {
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.id == expected_pid.id || ((tx.last_pid.id == expected_pid.id) && is_epoch_overflow(expected_pid.epoch));
```

Comment:
nit: Mind adding a comment when the second part of the condition is true? (for readers)